pkg/node-controller/controller.go

```diff
@@ -0,0 +1,676 @@
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	nad, err := c.nadLister.NetworkAttachmentDefinitions(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		// in this case the nad dne so it must've been deleted so we will cleanup nodeslicepools
+		// if we are down during the delete this could be missed similar to endpoints see kubernetes #6877
+		//TODO: do we want to cleanup IP Pools?
+		//err = c.whereaboutsclientset.WhereaboutsV1alpha1().NodeSlicePools(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+		//if err != nil && !errors.IsNotFound(err) {
+		//	return err
+		//}
+		//
+		//iPPools, err := c.ipPoolLister.IPPools(namespace).List(nil)
+		//if err != nil {
+		//	return err
+		//}
+		//for _, slicepool := range iPPools {
+		//	if ownerRef := metav1.GetControllerOf(slicepool); ownerRef != nil {
+		//		// If this object is not owned by a NodeSlicePool, we should not do anything more
+		//		// with it.
+		//		if ownerRef.Kind != "NodeSlicePool" || ownerRef.Name == name {
+		//			continue
+		//		}
+		//
+		//		err = c.whereaboutsclientset.WhereaboutsV1alpha1().IPPools(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+		//		if err != nil {
+		//			if !errors.IsNotFound(err) {
+		//				return err
+		//			}
+		//		}
+		//	}
+		//}
+		return nil
+	}
+	//nad does exist so did it change node_slice_range or slice_size
+	//TODO: error syncing 'default/igor-1': config file not found, requeuing
+	// can we make this work without a local config file
+	ipamConf, err := ipamConfiguration(nad, "")
+	if err != nil {
+		return err
+	}
+
+	// This is to support several NADs and interfaces on the same network
+	//TODO: validation that nad configs are compatible between different nads with same network-name
+	sliceName := name
+	if ipamConf.NetworkName != "" {
+		sliceName = ipamConf.NetworkName
+	}
+
+	logger.Info(fmt.Sprintf("%v", ipamConf))
+	if ipamConf.Range != "" {
+		logger.Info("skipping update node slices for network-attachment-definition due to range being set",
+			"network-attachment-definition", klog.KRef(namespace, name))
+		return nil
+	}
+	logger.Info(fmt.Sprintf("slicesize: %v", ipamConf.NodeSliceSize))
+	if ipamConf.NodeSliceFastRange == "" || ipamConf.NodeSliceSize == "" {
+		logger.Info("skipping update node slices for network-attachment-definition due missing node slice configurations",
+			"network-attachment-definition", klog.KRef(namespace, name))
+		return nil
+	}
+
+	logger.Info("About to update node slices for network-attachment-definition",
+		"network-attachment-definition", klog.KRef(namespace, name))
+
+	currentNodeSlicePool, err := c.nodeSlicePoolLister.NodeSlicePools(namespace).Get(sliceName)
+	if err != nil {
+		logger.Info("node slice pool does not exist, creating")
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		//Create
+		nodeslice := &v1alpha1.NodeSlicePool{
```

Comment:
We probably want to clear out the node slices on delete (or when all nads using the node slice are deleted), do we also want to clear out the IPPools?